src/metabase/lib/schema/literal.cljc

```diff
@@ -1,109 +1,118 @@
 (ns metabase.lib.schema.literal
   "Malli schemas for string, temporal, number, and boolean literals."
   (:require
-   [metabase.util.malli.registry :as mr]))
+   [malli.core :as mc]
+   [metabase.lib.schema.expression :as expression]
+   [metabase.util.malli.registry :as mr]
+   #?@(:clj ([metabase.lib.schema.literal.jvm]))))
+
+(defmethod expression/type-of* :dispatch-type/nil
+  [_nil]
+  :type/*)
+
+(mr/def ::boolean
+  :boolean)
+
+(defmethod expression/type-of* :dispatch-type/boolean
+  [_bool]
+  :type/Boolean)
 
 (mr/def ::boolean
   :boolean)
 
-;;; TODO -- for Clojure, we should also add BigInteger
 (mr/def ::integer
-  :int)
+  #?(:clj [:or
+           :int
+           :metabase.lib.schema.literal.jvm/big-integer]
+     :cljs :int))
+
+(defmethod expression/type-of* :dispatch-type/integer
+  [_int]
+  :type/Integer)
 
-;;; TODO -- for Clojure, we should also add BigDecimal, and `:float`:
-;;;
-;;; (malli.core/validate :double (float 1.0)) => false
-;;;
 ;;; we should probably also restrict this to disallow NaN and positive/negative infinity, I don't know in what
 ;;; universe we'd want to allow those if they're not disallowed already.
 (mr/def ::non-integer-real
-  :double)
+  #?(:clj [:or
+           :double
+           :metabase.lib.schema.literal.jvm/float
+           :metabase.lib.schema.literal.jvm/big-decimal]
+     :cljs :double))
+
+(defmethod expression/type-of* :dispatch-type/number
+  [_non-integer-real]
+  ;; `:type/Float` is the 'base type' of all non-integer real number types in [[metabase.types]] =(
+  :type/Float)
 
 (mr/def ::string
   :string)
 
 ;;; TODO -- these temporal literals could be a little stricter, right now they are pretty permissive, you shouldn't be
 ;;; allowed to have month `13` or `02-29` for example
 
-(def ^:private date-part-regex #"\d{4}-\d{2}-\d{2}")
-
-(def ^:private time-part-regex #"\d{2}:\d{2}(?::\d{2}(?:\.\d{3})?)?")
-
-(def ^:private offset-part-regex
-  ;; I think technically a zone offset can have a seconds part too but let's not worry too much about supporting that
-  ;; for now.
-  (re-pattern (str "(?:Z|" time-part-regex ")")))
+;;; TODO -- these were split out into separate parts originally but apparently string <-> re-pattern conversion
+;;; doesn't work in Cljs the way it works in Clj
```

Comment:
I think it would work if the fragments were written as strings (with the necessary extra escaping), concatenated, and then passed to `re-pattern`? But they're already written out and I'm not sure we really gain anything from the factoring.

Reply:
Hmmm, I didn't try that. I tried `(re-pattern (str ... <str> <regex> ...))` and that didn't work. But I didn't try with all strs